File: mac/OCRReview/Services/FindService.swift
import Foundation

struct FindMatch: Identifiable, Equatable {
    let id = UUID()
    let pageNumber: Int
    let range: Range<String.Index>
    let snippet: String
}

enum FindService {
    static func find(in document: OCRDocument, query: String, caseSensitive: Bool = false) -> [FindMatch] {
        let trimmed = query.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return [] }

        var matches: [FindMatch] = []
        for page in document.pages.sorted(by: { $0.pageNumber < $1.pageNumber }) {
            let text = page.displayText
            var searchStart = text.startIndex
            while searchStart < text.endIndex,
                  let range = text.range(
                      of: trimmed,
                      options: caseSensitive ? [] : .caseInsensitive,
                      range: searchStart..<text.endIndex
                  )
            {
                let snippet = snippet(for: text, around: range)
                matches.append(FindMatch(pageNumber: page.pageNumber, range: range, snippet: snippet))
                searchStart = range.upperBound
            }
        }
        return matches
    }

    static func replace(
        in text: String,
        query: String,
        replacement: String,
        at range: Range<String.Index>,
        caseSensitive: Bool = false
    ) -> String? {
        let matched = String(text[range])
        if caseSensitive {
            guard matched == query else { return nil }
        } else {
            guard matched.caseInsensitiveCompare(query) == .orderedSame else { return nil }
        }
        var result = text
        result.replaceSubrange(range, with: replacement)
        return result
    }

    static func replaceAll(in text: String, query: String, replacement: String, caseSensitive: Bool = false) -> String {
        let trimmed = query.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return text }
        var result = text
        var searchStart = result.startIndex
        while searchStart < result.endIndex,
              let range = result.range(
                  of: trimmed,
                  options: caseSensitive ? [] : .caseInsensitive,
                  range: searchStart..<result.endIndex
              )
        {
            result.replaceSubrange(range, with: replacement)
            searchStart = result.index(range.lowerBound, offsetBy: replacement.count)
        }
        return result
    }

    private static func snippet(for text: String, around range: Range<String.Index>, radius: Int = 24) -> String {
        let start = text.index(range.lowerBound, offsetBy: -radius, limitedBy: text.startIndex) ?? text.startIndex
        let end = text.index(range.upperBound, offsetBy: radius, limitedBy: text.endIndex) ?? text.endIndex
        var snippet = String(text[start..<end]).replacingOccurrences(of: "\n", with: " ")
        if start > text.startIndex { snippet = "…" + snippet }
        if end < text.endIndex { snippet += "…" }
        return snippet
    }
}

File: windows/src/OcrReview.Core/Services/FindService.cs
using System.Text;
using OcrReview.Core.Models;

namespace OcrReview.Core.Services;

public static class FindService
{
    public static List<FindMatch> Find(OcrDocument document, string query, bool caseSensitive = false)
    {
        var trimmed = query.Trim();
        var matches = new List<FindMatch>();
        if (trimmed.Length == 0) return matches;

        var cmp = caseSensitive ? StringComparison.Ordinal : StringComparison.OrdinalIgnoreCase;
        foreach (var page in document.Pages.OrderBy(p => p.PageNumber))
        {
            var text = page.DisplayText;
            int idx = 0;
            while (idx <= text.Length)
            {
                int found = text.IndexOf(trimmed, idx, cmp);
                if (found < 0) break;
                matches.Add(new FindMatch
                {
                    PageNumber = page.PageNumber,
                    Start = found,
                    Length = trimmed.Length,
                    Snippet = Snippet(text, found, trimmed.Length),
                });
                idx = found + Math.Max(trimmed.Length, 1);
            }
        }
        return matches;
    }

    public static string ReplaceAt(string text, int start, int length, string replacement)
    {
        if (start < 0 || start + length > text.Length) return text;
        return string.Concat(text.AsSpan(0, start), replacement, text.AsSpan(start + length));
    }

    public static string ReplaceAll(string text, string query, string replacement, bool caseSensitive = false)
    {
        var trimmed = query.Trim();
        if (trimmed.Length == 0) return text;

        var cmp = caseSensitive ? StringComparison.Ordinal : StringComparison.OrdinalIgnoreCase;
        var sb = new StringBuilder();
        int idx = 0;
        while (true)
        {
            int found = text.IndexOf(trimmed, idx, cmp);
            if (found < 0)
            {
                sb.Append(text, idx, text.Length - idx);
                break;
            }
            sb.Append(text, idx, found - idx);
            sb.Append(replacement);
            idx = found + trimmed.Length;
        }
        return sb.ToString();
    }

    private static string Snippet(string text, int start, int length, int radius = 24)
    {
        int from = Math.Max(0, start - radius);
        int toExclusive = Math.Min(text.Length, start + length + radius);
        var snippet = text.Substring(from, toExclusive - from).Replace('\n', ' ');
        if (from > 0) snippet = "…" + snippet;
        if (toExclusive < text.Length) snippet += "…";
        return snippet;
    }
}
